tests/framework/components/test_component.py

```diff
@@ -0,0 +1,252 @@
+    InteractiveContext(components=[ColumnCreator(), component])
+
+    # Assert population view is set and has the correct columns
+    assert component.population_view is not None
+    assert set(component.population_view.columns) == set(component.columns_required)
+
+
+def test_component_that_creates_and_requires_columns_population_view():
+    component = ColumnCreatorAndRequirer()
+    InteractiveContext(components=[ColumnCreator(), component])
+
+    # Assert population view is set and has the correct columns
+    expected_columns = component.columns_required + component.columns_created
+
+    assert component.population_view is not None
+    assert set(component.population_view.columns) == set(expected_columns)
+
+
+def test_component_that_requires_all_columns_population_view():
+    component = AllColumnsRequirer()
+    simulation = InteractiveContext(
+        components=[ColumnCreator(), ColumnCreatorAndRequirer(), component]
+    )
+    population = simulation.get_population()
+
+    # Assert population view is set and has the correct columns
+    expected_columns = population.columns
+
+    assert component.population_view is not None
+    assert set(component.population_view.columns) == set(expected_columns)
+
+
+def test_component_with_no_population_view():
+    component = NoPopulationView()
+    InteractiveContext(components=[ColumnCreator(), component])
+
+    # Assert population view is not set
+    assert component.population_view is None
+
+
+def test_component_initializer_is_not_registered_if_not_defined():
+    component = NoPopulationView()
+    simulation = InteractiveContext(components=[component])
+
+    # Assert that simulant initializer has been registered
+    assert component.on_initialize_simulants not in simulation._resource
+
+
+def test_component_initializer_is_registered_and_called_if_defined():
+    component = ColumnCreator()
+    simulation = InteractiveContext(components=[component])
+    population = simulation.get_population()
+    expected_pop_view = pd.DataFrame(
+        {column: 9 for column in component.columns_created}, index=population.index
+    )
+
+    # Assert that simulant initializer has been registered
+    assert component.on_initialize_simulants in simulation._resource
+    # and that created columns are correctly initialized
+    pd.testing.assert_frame_equal(population[component.columns_created], expected_pop_view)
+
+
+def test_listeners_are_not_registered_if_not_defined():
+    component = NoPopulationView()
+    simulation = InteractiveContext(components=[component])
+
+    time_step_prepare_methods = simulation.get_listeners("time_step__prepare")
+    time_step_methods = simulation.get_listeners("time_step")
+    time_step_cleanup_methods = simulation.get_listeners("time_step__cleanup")
+    collect_metrics_methods = simulation.get_listeners("collect_metrics")
+
+    for i in range(10):
+        assert component.on_time_step_prepare not in set(time_step_prepare_methods.get(i, []))
```

Comment:
Here and all the instances below I am a bit confused with how you know what integer key to call in the dict. Can you explain that a bit because I see above you did change the typing for the method to be a Dict[int, list]?

Reply:
The integer key here is the priority of the call. By default calls are registered to be at priority 5. Here I'm iterating through all the possible priorities to demonstrate that this call wasn't registered on any of them. Below, I know that the default priority is 5, so I'm checking there for `DefaultPriorities`, since that component doesn't specify a priority. For `CustomPriorities`, I specified those particular values, so I want to be sure that they have been registered in the right place.

See the comments in the definition of `register_listener` here fore more details: https://github.com/ihmeuw/vivarium/blob/main/src/vivarium/framework/event.py#L286

Reply:
Oh ok that makes sense.